### PR TITLE
Fix WebTorrent instant streaming playback links

### PR DESCRIPTION
**Pull request by feross**
In short, this PR does not redirect URL requests with `#ix=` to the .torrent file.

When the user clicks on a file in the torrent file list, a new tab opens which points to the .torrent file, but with a file fragment appended. For example, the torrent https://webtorrent.io/torrents/big-buck-bunny.torrent would have a link to https://webtorrent.io/torrents/big-buck-bunny.torrent#ix=1 for the 2nd file in the torrent.

Without this PR, this request passes the `IsWebtorrentInitiated(ctx)` condition and is therefore allowed to reach the network. Instead, we should confirm that the request is not for an individual file before letting it go to the network. That's what this PR does.

Fixes: https://github.com/brave/brave-browser/issues/3966

Branched from this PR: https://github.com/brave/brave-core/pull/2989 (so merge after that one is merged)

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- Go to https://webtorrent.io/free-torrents
- Click on Big Buck Bunny (torrent file)
- Start download
- Click on the .mp4 file (second file)
- Confirm that it opens a new tab, and a video player shows up.
- Confirm that the video plays Big Buck Bunny when it is clicked.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper.cc [hunk 1]
```diff
@@ -41,6 +41,16 @@ bool URLMatched(const GURL& url) {
...
+/**
+ * Returns true if the URL contains a URL fragment that starts with "ix=". For
+ * example, https://webtorrent.io/torrents/big-buck-bunny.torrent#ix=1.
+ * Otherwise, returns false.
+ */
+bool IsViewerURL(const GURL& url) {
+  return base::StartsWith(url.ref(), "ix=",
+      base::CompareCase::INSENSITIVE_ASCII);
+}
+
```
Comment (yrliou) [inline components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper.cc]: Could you add a test case for this check in https://github.com/brave/brave-core/blob/e3f81b02176d9b9bf4e61efda215b50b0fa4d111/components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper_unittest.cc?

Comment (feross) [reply components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper.cc]: Added the test, as you requested.

File: components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper_unittest.cc [hunk 2]
```diff
@@ -37,10 +37,16 @@ class BraveTorrentRedirectNetworkDelegateHelperTest: public testing::Test {
     context_->Init();
 
     torrent_url_ = GURL("https://webtorrent.io/torrents/sintel.torrent");
+    torrent_viewer_url_ =
+      GURL("https://webtorrent.io/torrents/sintel.torrent#ix=0");
     non_torrent_url_ = GURL("https://webtorrent.io/torrents/sintel");
-    extension_url_ = GURL(
```
Comment (feross) [inline components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper_unittest.cc]: I renamed the `extension_url_` variable to `torrent_extension_url_` so the naming is consistent across all variables.

**Comment by feross** [comment]
Sure thing. Looks like I got unlucky and the version just changed :/

**Comment by yrliou** [comment]
@feross Please help fix lint errors below.
```
14:31:59  components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper_unittest.cc:49:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
14:31:59  components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper_unittest.cc:333:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
14:31:59  components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper_unittest.cc:370:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
```
You may run `npm run lint` under brave-browser locally to check the result too.

**Comment by feross** [comment]
> You may run `npm run lint` under brave-browser locally to check the result too.

Thanks - didn't realize that there was an `npm run lint` for `brave-browser` that differed from the one under `brave-core`. Good to know, thanks!

**Comment by yrliou** [comment]
```
18:51:19  [227/227] BraveRewardsBrowserTest.ProcessPendingContributions (TIMED OUT)
18:51:19  1 test timed out:
18:51:19      BraveRewardsBrowserTest.ProcessPendingContributions (../../brave/components/brave_rewards/browser/rewards_service_browsertest.cc:2153)
```
Above was the test failure, which isn't relevant to this PR.

File: components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper.cc [hunk 3]
```diff
@@ -41,6 +41,16 @@ bool URLMatched(const GURL& url) {
       base::CompareCase::INSENSITIVE_ASCII);
 }
 
+/**
+ * Returns true if the URL contains a URL fragment that starts with "ix=". For
```
Comment (iefremov) [inline components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper.cc]: @feross Just some nits - we use google/chromium style guide that requires `//`-style comments

File: components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper_unittest.cc [hunk 4]
```diff
@@ -341,3 +361,44 @@ TEST_F(BraveTorrentRedirectNetworkDelegateHelperTest,
...
+       WebtorrentInitiatedViewerURLRedirect) {
+  net::TestDelegate test_delegate;
+  std::unique_ptr<net::URLRequest> request =
+      context()->CreateRequest(torrent_viewer_url(), net::IDLE, &test_delegate,
+                               TRAFFIC_ANNOTATION_FOR_TESTS);
+
+  request->set_initiator(url::Origin::Create(
+    torrent_extension_url().GetOrigin()));
+  scoped_refptr<net::HttpResponseHeaders> orig_response_headers =
+    new net::HttpResponseHeaders(std::string());
```
Comment (iefremov) [inline components/brave_webtorrent/browser/net/brave_torrent_redirect_network_delegate_helper_unittest.cc]: it is handy to use `clang-format` to make whitespaces CC-conformant

**Comment by feross** [comment]
@iefremov Lmk if you want me to send a PR to fix these issues.

**Pull request merged**